### PR TITLE
Changes from CoopSymbiotic

### PR DESCRIPTION
**Pull request by JoeMurray**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 *~
 *.bak
+civicrm_constants.local.php
```
Comment (JoeMurray) [inline .gitignore]: Okay, I probably should have been more selective as I created my PR. Obviously this should not be merged.

File: CRM/Cdntaxcalculator/Page/Province.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+class CRM_Cdntaxcalculator_Page_Province extends CRM_Core_Page {
+
+  function run() {
+    $result = [
+      'status' => 0,
+    ];
+
+    $province_id = intval($_REQUEST['state_province_id']);
```
Comment (pradpnayak) [inline CRM/Cdntaxcalculator/Page/Province.php]: Please use CRM_Utils_Request::retrieve('state_province_id', 'Integer', $this);

File: CRM/Cdntaxcalculator/BAO/CDNTaxes.php [hunk 3]
```diff
@@ -1,58 +1,362 @@
...
+        }
+      }
+    }
+
+    if (empty($province_id)) {
+      $domain_id = CRM_Core_Config::domainID();
+
+      $result = civicrm_api3('Domain', 'getsingle', [
+        'id' => $domainID,
+      ]);
```
Comment (pradpnayak) [inline CRM/Cdntaxcalculator/BAO/CDNTaxes.php]: This will throw exception incase of multi-domain.

File: CRM/Cdntaxcalculator/BAO/CDNTaxes.php [hunk 4]
```diff
@@ -1,58 +1,362 @@
...
+
+      $province_id = $result['domain_address']['state_province_id'];
+      $country_id = $result['domain_address']['country_id'];
+    }
+
+    if (empty($province_id) || empty($country_id)) {
+      CRM_Core_Error::fatal("Failed to find a default country/province for the event.");
+    }
+
+    if ($country_id == 1039) {
```
Comment (pradpnayak) [inline CRM/Cdntaxcalculator/BAO/CDNTaxes.php]: Hard coded :(

File: CRM/Cdntaxcalculator/BAO/CDNTaxes.php [hunk 5]
```diff
@@ -1,58 +1,362 @@
...
+    $is_taxed = CRM_Core_DAO::singleValueQuery('
+      SELECT count(*)
+        FROM civicrm_entity_financial_account efa
+        LEFT JOIN civicrm_financial_account fa on fa.id = efa.financial_account_id
+       WHERE efa.entity_id = 9
+         AND efa.entity_table = "civicrm_financial_type"
+         AND fa.is_tax = 1', [
+      1 => [$event_id, 'Positive'],
+    ]);
+
```
Comment (pradpnayak) [inline CRM/Cdntaxcalculator/BAO/CDNTaxes.php]: This is not always true since an event might be configured to have priceset having some price fields taxable. So it might be possible that event.financial_type_id is not taxable.

File: cdntaxcalculator.php [hunk 6]
```diff
@@ -108,32 +108,140 @@ function cdntaxcalculator_civicrm_alterSettingsFolders(&$metaDataFolders = NULL)
...
+  elseif ($pageType == 'event') {
+    $event_id = $form->get('id');
+    $taxes = CRM_Cdntaxcalculator_BAO_CDNTaxes::getTaxesForEvent($event_id);
+    $province_id = $taxes['province_id'];
+  }
+  else {
+    // Country/Province ID was passed an an URL argument
+    // ex: redirecting after selecting a province from the popup.
+    if (!empty($_GET['cdntax_country_id'])) {
+      $country_id = intval($_GET['cdntax_country_id']);
```
Comment (pradpnayak) [inline cdntaxcalculator.php]: Please use CRM_Utils_Request::retrieve('cdntax_country_id', 'Integer', $this);

File: cdntaxcalculator.php [hunk 7]
```diff
@@ -108,32 +108,140 @@ function cdntaxcalculator_civicrm_alterSettingsFolders(&$metaDataFolders = NULL)
...
+    }
+
+    if (!empty($_GET['cdntax_province_id'])) {
+      $province_id = intval($_GET['cdntax_province_id']);
```
Comment (pradpnayak) [inline cdntaxcalculator.php]: same as above

File: cdntaxcalculator.php [hunk 8]
```diff
@@ -108,32 +108,140 @@ function cdntaxcalculator_civicrm_alterSettingsFolders(&$metaDataFolders = NULL)
...
+
+    // The user is logged-in.
+    if (empty($province_id) && !empty($contact_id)) {
+      $province_id = cdn_getStateProvince($contact_id);
+    }
+
+    // FIXME: when is this used?
+    // FIXME: potential info leak if we let users lookup provinces of any contact?
+    if (empty($province_id) && empty($contact_id) && !empty($_GET['contactId'])) {
+      $contact_id = $_GET['contactId'];
```
Comment (pradpnayak) [inline cdntaxcalculator.php]: same as above

File: CRM/Cdntaxcalculator/BAO/CDNTaxes.php [hunk 9]
```diff
@@ -20,30 +22,36 @@ static public function applyTaxesToPriceset(&$feeBlock, &$taxes) {
           $has_taxable_amounts = TRUE;
         }
       }
+
+      CRM_Utils_Hook::singleton()->invoke(['line_items'], $fee['options'], $null, $null, $null, $null, $null, 'cdntaxcalculator_alter_lineitems');
```
Comment (pradpnayak) [inline CRM/Cdntaxcalculator/BAO/CDNTaxes.php]: This is not a standard way to define hooks or invoke hook calls.  CRM_Utils_Hook_CDN Create a class that will extend CRM_Utils_Hook and then define hook and invoke  the hook call here CRM_Utils_Hook_CDN::alterLineItems();

**Comment by pradpnayak** [comment]
I did some QA and the PR works fine. Total tax is calculated properly according to Canadian Province. However its unclear when viewing a contribution to determine how much HST and/or PST is charged for line item.

Happy to merge the PR if the changes requested are fixed.

**Comment by mlutfy** [comment]
Thanks for the reviews, @pradpnayak ! Tax separation (GST/HST and PST) is also high on our list. Another high priority is recalculation of arbitrary (not pricesets) taxable amounts.

> "However its unclear when viewing a contribution to determine how much HST and/or PST is charged for line item."

The extension mostly changes tax rates on line items (for a given province), so this would be a civicrm core limitation?

capture d ecran de 2017-11-28 14-56-36

**Comment by monishdeb** [comment]
@JoeMurray @mlutfy I will be looking into this PR and will create separate sub-PRs of whats needed.

**Comment by mlutfy** [comment]
@monishdeb We're still experimenting/changing things a lot, and some of the work requires a patch on core ([diff with debug statements, needs cleanup](https://gist.github.com/mlutfy/3c3d74d24c5853bf39d29830bc8e93d8)).

Unless you need this extension for a project, it might be better to wait a bit, until things stabilise a bit?

There are a few core tax open PRs, however:

* https://github.com/civicrm/civicrm-core/pull/11459 (reviewed by Jitendra, but I haven't had time to test/review)
* https://github.com/civicrm/civicrm-core/pull/11461
* https://github.com/civicrm/civicrm-core/pull/11513 (sounds related to 11461, but not sure if the same)

:-)

**Comment by monishdeb** [comment]
@mlutfy no problem. Please ping me when you want me to review this patch(and other PRs if any) so that I can pull these fixes in extension.

As per the status of those 3 core PRs:
1. https://github.com/civicrm/civicrm-core/pull/11459 (merged)
2. https://github.com/civicrm/civicrm-core/pull/11461 (although not required, as the issue was fixed in https://github.com/civicrm/civicrm-core/pull/11455. Still, we have asked Mattwire if there are any circumstances left to fix)
3. https://github.com/civicrm/civicrm-core/pull/11513 (Replaced with https://github.com/civicrm/civicrm-core/pull/11521 and added UT :) )

File: .gitignore (hunk 1 above)
Comment (JoeMurray) [reply .gitignore]: @monishdeb Could you review in the next couple of days and merge as appropriate. Thanks, @mlutfy !!

**Comment by mlutfy** [comment]
@JoeMurray  Sorry, I had forgotten this PR was still opened. Thanks for the review so far! I still need to follow-up on some items.

We have been doing a lot of changes, supporting more use-cases, but it still requires some patches on core, which we have not well documented yet.

Before promoting this more broadly, I also want to start working on tests, e2e and with phpunit, but rather far from there at the moment (we only have some very basic tests with puppeteer, and it's running against one of our pre-configured test sites).

In short: I guess I want to promote this to people who need a tax solution (I still stumble on integrations that do ad-hoc JS hacks), but it still needs a fair amount of work/tests/documentation for it to be more robust.

File: .gitignore (hunk 1 above)
Comment (mlutfy) [reply .gitignore]: To clarify: `civicrm_constants.local.php` is a new file that makes it possible to override defaults in `civicrm_constants.php` (to change tax names / to translate them).
